diff --git a/include/fuse_common.h b/include/fuse_common.h
index beb44c7d09899da4d768e8b31df7928df95e3672..b0fcb0464bdbcaa8b1d73186dd763015e98483da 100644
--- a/include/fuse_common.h
+++ b/include/fuse_common.h
@@ -102,6 +102,7 @@ struct fuse_file_info {
  * FUSE_CAP_ASYNC_DIO: asynchronous direct I/O submission
  * FUSE_CAP_WRITEBACK_CACHE: use writeback cache for buffered writes
  * FUSE_CAP_NO_OPEN_SUPPORT: support zero-message opens
+ * FUSE_CAP_EXTFUSE: support ExtFUSE bpf extensions
  */
 #define FUSE_CAP_ASYNC_READ		(1 << 0)
 #define FUSE_CAP_POSIX_LOCKS		(1 << 1)
@@ -120,6 +121,7 @@ struct fuse_file_info {
 #define FUSE_CAP_ASYNC_DIO		(1 << 15)
 #define FUSE_CAP_WRITEBACK_CACHE	(1 << 16)
 #define FUSE_CAP_NO_OPEN_SUPPORT	(1 << 17)
+#define FUSE_CAP_EXTFUSE			(1 << 26)
 
 /**
  * Ioctl flags
@@ -199,10 +201,13 @@ struct fuse_conn_info {
 	 */
 	unsigned time_gran;
 
+	/* ExtFUSE prog */
+	unsigned extfuse_prog_fd;
+
 	/**
 	 * For future use.
 	 */
-	unsigned reserved[22];
+	unsigned reserved[21];
 };
 
 struct fuse_session;
diff --git a/include/fuse_kernel.h b/include/fuse_kernel.h
index c9aca042e61d197927409531af4f8cdd88218adc..4f7d635447b7e0705c6a5fde04368bb772e55af8 100644
--- a/include/fuse_kernel.h
+++ b/include/fuse_kernel.h
@@ -231,6 +231,7 @@ struct fuse_file_lock {
  * FUSE_ASYNC_DIO: asynchronous direct I/O submission
  * FUSE_WRITEBACK_CACHE: use writeback cache for buffered writes
  * FUSE_NO_OPEN_SUPPORT: kernel supports zero-message opens
+ * FUSE_FS_EXTFUSE: enable ExtFUSE bpf extensions
  */
 #define FUSE_ASYNC_READ		(1 << 0)
 #define FUSE_POSIX_LOCKS	(1 << 1)
@@ -250,6 +251,7 @@ struct fuse_file_lock {
 #define FUSE_ASYNC_DIO		(1 << 15)
 #define FUSE_WRITEBACK_CACHE	(1 << 16)
 #define FUSE_NO_OPEN_SUPPORT	(1 << 17)
+#define FUSE_FS_EXTFUSE		(1 << 26)
 
 /**
  * CUSE INIT request/reply flags
@@ -587,7 +589,9 @@ struct fuse_init_out {
 	uint16_t	congestion_threshold;
 	uint32_t	max_write;
 	uint32_t	time_gran;
-	uint32_t	unused[9];
+	uint32_t	padding;
+	uint32_t	extfuse_prog_fd;
+	uint32_t	unused[7];
 };
 
 #define CUSE_INIT_INFO_MAX 4096
diff --git a/lib/fuse_lowlevel.c b/lib/fuse_lowlevel.c
index c5108f73858bbb7b626ca4584940c9ef7e43645f..1bff11872c2fdce6fefdb58fd3a3d263c7d59b4b 100755
--- a/lib/fuse_lowlevel.c
+++ b/lib/fuse_lowlevel.c
@@ -1949,6 +1949,15 @@ static void do_init(fuse_req_t req, fuse_ino_t nodeid, const void *inarg)
 	if (req->f->conn.proto_minor >= 18)
 		f->conn.capable |= FUSE_CAP_IOCTL_DIR;
 
+	if (arg->flags & FUSE_FS_EXTFUSE) {
+		if (f->debug)
+			fprintf(stderr, "Enabling FUSE_FS_EXTFUSE capability\n");
+		f->conn.capable |= FUSE_CAP_EXTFUSE;
+	} else {
+		if (f->debug)
+			fprintf(stderr, "FUSE_FS_EXTFUSE capability not present!\n");
+	}
+
 	if (f->atomic_o_trunc)
 		f->conn.want |= FUSE_CAP_ATOMIC_O_TRUNC;
 	if (f->op.getlk && f->op.setlk && !f->no_remote_posix_lock)
@@ -2024,6 +2033,13 @@ static void do_init(fuse_req_t req, fuse_ino_t nodeid, const void *inarg)
 		outarg.flags |= FUSE_ASYNC_DIO;
 	if (f->conn.want & FUSE_CAP_WRITEBACK_CACHE)
 		outarg.flags |= FUSE_WRITEBACK_CACHE;
+	if (f->conn.want & FUSE_CAP_EXTFUSE) {
+		outarg.flags |= FUSE_FS_EXTFUSE;
+		outarg.extfuse_prog_fd = f->conn.extfuse_prog_fd;
+		if (f->debug)
+			fprintf(stderr, "   ExtFUSE prog fd: %d\n",
+				outarg.extfuse_prog_fd);
+	}
 	outarg.max_readahead = f->conn.max_readahead;
 	outarg.max_write = f->conn.max_write;
 	if (f->conn.proto_minor >= 13) {
